### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by Aoxian**
I've implemented all the suggestions from Brian, updated tests and added IO tests using Capture IO.  This includes game play throughs. If all looks good, I can get back to computer ai stuff.

### CODE REVIEW

File: lib/game_loop.ex [hunk 1]
```diff
@@ -100,6 +100,6 @@ defmodule GameLoop do
   end
 
   def process_play_again(play_again) do
-    if play_again == "y", do: Game.start(), else: {:ok, :game_over}
+    if play_again == "y", do: TicTacToe.start(), else: {:ok, :game_over}
```
Comment (nystromb) [inline lib/game_loop.ex]: Why call `TicTacToe` here? Now theres a co dependency on `TicTacToe` and `GameLoop`

Why not just call the `run` function here? I dont think `GameLoop` should know about the main module.

File: lib/game_loop.ex [hunk 2]
```diff
@@ -73,7 +73,7 @@ defmodule GameLoop do
   end
 
   def check_for_end_game(state) do
-    case Game.check_status(state.board) do
+    case GameStatus.check(state.board) do
```
Comment (nystromb) [inline lib/game_loop.ex]: I might rename this to `game_over?` or just `over?` but the function name should describe what it's doing. Checking what...?

File: lib/game_status.ex [hunk 3]
```diff
@@ -0,0 +1,19 @@
+defmodule GameStatus do
```
Comment (nystromb) [inline lib/game_status.ex]: 👍

File: lib/game_loop.ex [hunk 4]
```diff
@@ -2,78 +2,44 @@ defmodule GameLoop do
...
     |> get_next_player()
     |> run()
   end
 
-  def clear_screen(state) do
-    GameUI.clear_screen(state.ui)
+  defp refresh_display(%{board: board, message: message, ui: ui} = state) do
+    ui.clear_screen()
+    ui.display_board(board)
+    ui.display_message(message)
```
Comment (nystromb) [inline lib/game_loop.ex]: Have you thought about moving this function into the console UI? I feel like you can have a single function `display_board`, which will:

1. clear the screen
2. display the board
3. display the message (without telling it the message you want to send) 

You should be able to just pass it the state (the board, the atom of where the status is at, and the `ConsoleUI` should know what to print based off that.

File: lib/game_loop.ex (hunk 1 above)
Comment (Aoxian) [reply lib/game_loop.ex]: I move the initial beginning state for a console game out to the TicTacToe module at Justin's suggestion, hence why it is calling out.  I'll have to think about this one.  Is there a place where beginning state are typically kept in functional languages?

File: lib/game_loop.ex [hunk 5]
```diff
@@ -2,78 +2,44 @@ defmodule GameLoop do
...
-            | move: :invalid,
-              message: "That is not an available square, please try again"
-          }
-
-      {:error, :invalid_key} ->
-        %{state | move: :invalid, message: "Did not receive a valid move, please try again"}
-    end
+  def get_move(%{ui: ui} = state) do
+    ui.get_user_input("Choose a spot [1-9] or q to quit\n")
+    |> Game.move(state)
```
Comment (nystromb) [inline lib/game_loop.ex]: This is a Single Responsibility Principle violation. This not only gets the user input, but also places the the input on the board. You can probably separate these out in your run function easily.

File: test/game_loop_test.exs [hunk 6]
```diff
@@ -1,5 +1,40 @@
...
+  describe "TTT play through scenarios" do
+    test "The game can end in a tie" do
+      assert capture_io(
+               [input: "1\n2\n3\n7\n8\n9\n4\n5\n6\nq\n"],
+               fn -> TicTacToe.start() end
+             ) =~ "Meh. Tie game."
+    end
+
+    test "X can win the game" do
+      assert capture_io(
```
Comment (nystromb) [inline test/game_loop_test.exs]: 💯 👍 💪

File: lib/game_loop.ex [hunk 7]
```diff
@@ -93,13 +59,9 @@ defmodule GameLoop do
...
-      GameUI.get_user_input(state.ui, "Type 'y' to play again, anything else to quit")
-    )
-  end
-
-  def process_play_again(play_again) do
-    if play_again == "y", do: Game.start(), else: {:ok, :game_over}
+  def ask_to_play_again(%{ui: ui}) do
+    if ui.get_user_input("Type 'y' to play again, anything else to quit") == "y",
+      do: TicTacToe.start(),
+      else: {:ok, :game_over}
```
Comment (nystromb) [inline lib/game_loop.ex]: I'm not sure if passing back `:ok` as an atom is really helpful, and it appears you aren't using it at all.

I might instead get rid of the `:ok` and instead return `{:game_over, state}`. The atom is really just used for pattern matching.

File: lib/game_loop.ex (hunk 1 above)
Comment (Aoxian) [reply lib/game_loop.ex]: I ended up putting it in Game for now.  Dependency, destroyed.

File: lib/game_loop.ex (hunk 7 above)
Comment (Aoxian) [reply lib/game_loop.ex]: Yeah, I don't know why I did that.  I like your suggestion of {:game_over, state} and updated the code.

File: lib/game_loop.ex (hunk 2 above)
Comment (Aoxian) [reply lib/game_loop.ex]: I renamed the function to game_over? as it does check for that.

For some reason I thought '?'s in method names were reserved for those hat return a boolean, this function returns a tuple with a result (ok or error) and a message (x_wins, o_wins, tie_game, in_progress or invalid_mark).

File: lib/game_loop.ex (hunk 4 above)
Comment (Aoxian) [reply lib/game_loop.ex]: I need to break it out of the pipeline to do this, as I don't want the ConsuleUI to know about the game state.  Code gets ugly, but I can do it.

Comment (Aoxian) [reply lib/game_loop.ex]: The message is literally every message I set throughout the code, so I need to send the message as well as the board.

File: lib/game_loop.ex (hunk 5 above)
Comment (Aoxian) [reply lib/game_loop.ex]: I resolved  the issue. Making sure the UI is not aware of game state means I had to pull this out of the pipeline.  I could have a little wrapper, but I didn't think it was necessary.

File: lib/game_loop.ex (hunk 4 above)
Comment (nystromb) [reply lib/game_loop.ex]: @Aoxian why shouldn't the ConsoleUI know about the game state?

File: lib/game_loop.ex (hunk 2 above)
Comment (nystromb) [reply lib/game_loop.ex]: You are right, i totally forgot its not returning a boolean. I might remove that then. 

https://github.com/elixir-lang/elixir/blob/master/lib/elixir/pages/Naming%20Conventions.md#trailing-question-mark-foo

File: lib/game_loop.ex (hunk 4 above)
Comment (Aoxian) [reply lib/game_loop.ex]: I thought it wasn't supposed to; the UI used to know about the state.  If I put it back it will be easier.  Let me try that.

Comment (nystromb) [reply lib/game_loop.ex]: I think the ConsoleUI should know about the state of the game, because that is how it will know what to print out based on that.

Comment (nystromb) [reply lib/game_loop.ex]: @td-extreme @7hoenix thoughts?

Comment (Aoxian) [reply lib/game_loop.ex]: I ended up moving it back out.  I like it.  The UI knows about the state, but I guess that makes sense as displaying the state to the user is part of what it does.

File: lib/game_loop.ex (hunk 2 above)
Comment (Aoxian) [reply lib/game_loop.ex]: game_over?() is now game_over()

**Comment by 7hoenix** [comment]
This is great. Nice job @Aoxian

**Pull request merged**